l1-contracts/contracts/bridge/L1Nullifier.sol

```diff
@@ -0,0 +1,733 @@
+            assembly {
+                callSuccess := call(gas(), ntvAddress, amount, 0, 0, 0, 0)
+            }
+            require(callSuccess, "ShB: eth transfer failed");
+        } else {
+            IERC20(_token).safeTransfer(ntvAddress, IERC20(_token).balanceOf(address(this)));
+        }
+    }
+
+    /// @notice Clears chain balance for specific token.
+    /// @dev This function is part of the upgrade process used to nullify chain balances once they are credited to NTV.
+    /// @param _chainId The ID of the ZK chain.
+    /// @param _token The address of the token which was previously deposit to shared bridge.
+    function clearChainBalance(uint256 _chainId, address _token) external {
+        require(msg.sender == address(nativeTokenVault), "ShB: not NTV");
+        chainBalance[_chainId][_token] = 0;
+    }
+
+    /// @notice Sets the L1ERC20Bridge contract address.
+
+    /// @dev Should be called only once by the owner.
+    /// @param _legacyBridge The address of the legacy bridge.
+    function setL1Erc20Bridge(address _legacyBridge) external onlyOwner {
+        require(address(legacyBridge) == address(0), "ShB: legacy bridge already set");
+        require(_legacyBridge != address(0), "ShB: legacy bridge 0");
+        legacyBridge = IL1ERC20Bridge(_legacyBridge);
+    }
+
+    /// @notice Sets the L1ERC20Bridge contract address.
+    /// @dev Should be called only once by the owner.
+    /// @param _nativeTokenVault The address of the native token vault.
+    function setNativeTokenVault(IL1NativeTokenVault _nativeTokenVault) external onlyOwner {
+        require(address(nativeTokenVault) == address(0), "Nullifier: native token vault already set");
+        require(address(_nativeTokenVault) != address(0), "Nullifier: native token vault 0");
+        nativeTokenVault = _nativeTokenVault;
+    }
+
+    /// @notice Sets the L1 asset router contract address.
+    /// @dev Should be called only once by the owner.
+    /// @param _l1AssetRouter The address of the native token vault.
+    function setL1AssetRouter(IL1AssetRouter _l1AssetRouter) external onlyOwner {
+        require(address(nativeTokenVault) == address(0), "Nullifier: l1 asset router already set");
+        require(address(_l1AssetRouter) != address(0), "Nullifier: l1 asset router 0");
+        l1AssetRouter = _l1AssetRouter;
+    }
+
+    /// @notice Confirms the acceptance of a transaction by the Mailbox, as part of the L2 transaction process within Bridgehub.
+    /// This function is utilized by `requestL2TransactionTwoBridges` to validate the execution of a transaction.
+    /// @param _chainId The chain ID of the ZK chain to which confirm the deposit.
+    /// @param _txDataHash The keccak256 hash of 0x01 || abi.encode(bytes32, bytes) to identify deposits.
+    /// @param _txHash The hash of the L1->L2 transaction to confirm the deposit.
+    function bridgehubConfirmL2Transaction(
```

Comment:
This is an unsolved issue, if we route all confirmations here it will not be backward compatible... Maybe send it through the AssetRouter?

Reply:
But we change the bridgehub and set nullifier there.
And the method can only be called by bridgehub.